1668 Sutarciu valdymas

martynasracct opened the pull request:
**Summary:**
* Adds `Agreement` and `AgreementScope` models.
* Adds endpoint `projects/{project_id}/agreement/add/` and HTML to create `Agreement` and `AgreementScope` objects.
    * `Agreement` object is created for each unique organization in `Project.datasets` (a.k.a use case datasets).
    * It's now possible to choose between 3 hardcoded scopes for each dataset: `<dataset_name>_getall`, `<dataset_name>_search`, `<dataset_name>_select` `AgreementScope`. 
* Button that redirects to Agreement creation will be added in different PR. Button will be added to Agreement list page.

**Ticket:**
https://github.com/atviriduomenys/katalogas/issues/1668

martynasracct committed:
1688 Pridėti sutarčių modeliai ir Django admin

martynasracct committed:
1688 Endpoint for creating Agreement with AgreementScopes

karina-klinkeviciute commented on vitrina/smart_contracts/models.py:
```diff
@@ -19,3 +20,74 @@ class SmartContractTemplate(UUIDBaseModel):
+    ACTIVE = "ACTIVE"
+    DEACTIVATED = "DEACTIVATED"
+
+    STATUSES = (
```
optional: textchoices can be used: https://docs.djangoproject.com/en/5.2/ref/models/fields/#enumeration-types

karina-klinkeviciute commented on vitrina/smart_contracts/models.py:
```diff
@@ -19,3 +20,74 @@ class SmartContractTemplate(UUIDBaseModel):
+
+    use_case = models.ForeignKey(
+        Project,
+        related_name="agreement_set",
```
Šita eilutė neturi prasmės, nebent `related_name` pervadintume į `agreements` ar panašiai - bet by default jis jau yra `agreement_set`

karina-klinkeviciute commented on vitrina/smart_contracts/models.py:
```diff
@@ -19,3 +20,74 @@ class SmartContractTemplate(UUIDBaseModel):
+        (DEACTIVATED, _("Nutraukta")),
+    )
+
+    use_case = models.ForeignKey(
```
optional: ar ir kitur panaudojimo atvejis naudojamas use_case, ar gal geriau nuosekliai naudoti `project?

Struchid commented on vitrina/locale/en/LC_MESSAGES/django.po:
```diff
@@ -2890,6 +2890,60 @@ msgstr "Page without side menu"
+msgstr "Active"
+
+msgid "Nutraukta"
+msgstr "Deactivated"
```
This seems to be a better translation for when it comes to contracts IMO:

```suggestion
msgstr "Terminated"
```

Struchid commented on vitrina/locale/en/LC_MESSAGES/django.po:
```diff
@@ -2890,6 +2890,60 @@ msgstr "Page without side menu"
+msgid "Nutraukta"
+msgstr "Deactivated"
+
+msgid "Statusas"
```
I suggest, since "Statusas" is not as Lithuanian as i'd like it to be, but that's only my opinion:D

```suggestion
msgid "Būsena"
```

or

```suggestion
msgid "Būklė"
```

Struchid commented on vitrina/locale/en/LC_MESSAGES/django.po:
```diff
@@ -2890,6 +2890,60 @@ msgstr "Page without side menu"
+msgid "Sutartis sėkmingai sugeneruota"
+msgstr "Agreement generated successfully"
+
+msgid "Sutartis generavime kilo klaidų"
```
```suggestion
msgid "Sutarties generavime kilo klaidų"
```

Struchid commented on tests/smart_contracts/test_views.py:
```diff
@@ -0,0 +1,165 @@
+        response = app.get(
+            reverse("agreement-create", args=[project.pk]), expect_errors=True
+        )
+        assert response.status_code == 403
```
Is the `response.json` formatted nicely or just a plain old error? If yes, we might want to check the return message as well

martynasracct replied:
It's `html/text` error sadly

Struchid commented on vitrina/smart_contracts/models.py:
```diff
@@ -19,3 +21,56 @@ class SmartContractTemplate(UUIDBaseModel):
+
+    class Meta:
+        managed = True
+        db_table = "usecase_agreement"
```
Do we really want a specific table name? This seems to cause confusion among our friends and I would avoid it if possible

martynasracct replied:
This model's main relation is with `Project` model. `Project` model has db_name="usecase". So I think it's better to keep it consistent at least in DB

karina-klinkeviciute replied:
db_name = usecase is for historical reasons, it's because the tables where created before Django project. I think that the default table name (agreement) is ok

martynasracct replied:
Removed custom DB table name

Struchid commented on vitrina/smart_contracts/models.py:
```diff
@@ -19,3 +21,56 @@ class SmartContractTemplate(UUIDBaseModel):
+        verbose_name=_("Statusas"),
+    )
+
+    agent_sync_enabled = models.BooleanField(
```
**Optional:** But I would love you for this:

```suggestion
    is_agent_sync_enabled = models.BooleanField(
```

Struchid commented on vitrina/smart_contracts/forms.py:
```diff
@@ -0,0 +1,48 @@
+                choice_name = dataset_metadata.name.replace("/", "_")
+                choices.append((f"{choice_name}_getall", f"{choice_name}_getall"))
+                choices.append((f"{choice_name}_search", f"{choice_name}_search"))
+                choices.append((f"{choice_name}_select", f"{choice_name}_select"))
```
**Optional:**

```suggestion
                choices.extend([
                    f"{choice_name}_getall", f"{choice_name}_getall",
                    f"{choice_name}_search", f"{choice_name}_search",
                    f"{choice_name}_select", f"{choice_name}_select"
                ])
```

Struchid commented on vitrina/smart_contracts/forms.py:
```diff
@@ -0,0 +1,48 @@
+            datasets_by_organization.get(self.instance, [])
+        )
+
+    scopes = forms.MultipleChoiceField(
```
I'd definitely put field definitions at the top of the class

Struchid commented on vitrina/smart_contracts/views.py:
```diff
@@ -0,0 +1,124 @@
+
+    def get_formset(self) -> BaseFormSet:
+        formset_kwargs = self.get_formset_kwargs()
+        datasets = self.object.datasets.all().order_by("organization_id")
```
I might see a catch here: do we get the `deleted=True` datasets here as well? Please check the other queries as well and their models, if they're hiding one of those fields.

varzgalys replied:
If true, deleted datasets should be excluded inside Model's manager (custom manager)

martynasracct replied:
Yeep. Deleted datasets are excluded in custom manager. Changed this code a bit: now `public_datasets` are queried using `Prefetch()` on `Project` query.

varzgalys commented on vitrina/smart_contracts/forms.py:
```diff
@@ -0,0 +1,48 @@
+    @staticmethod
+    def create_scope_choices(datasets: list[Dataset]) -> list[tuple[str, str]]:
+        choices = []
+        for dataset in datasets:
+            if dataset_metadata := dataset.metadata.first():
```
If many datasets - performance wise to collect all metadata in one query, by:

1. Using prefetch_realted during select before this method.
or
2. Metadata.objects.filter(dataset__in=datasets)

varzgalys replied:
Comment optional

martynasracct replied:
Do you know if `prefetch_related` will save a query even when calling `.first()`?

varzgalys commented on vitrina/smart_contracts/views.py:
```diff
@@ -0,0 +1,124 @@
+        datasets = self.object.datasets.all().order_by("organization_id")
+        datasets_by_organization = {
+            organization: list(datasets)
+            for organization, datasets in groupby(datasets, lambda d: d.organization)
```
```suggestion
            organization_id: list(datasets)
            for organization_id, datasets in groupby(datasets, lambda d: d.organization)
```

Right?

martynasracct replied:
No, the grouper is organization object. Not `id`

varzgalys commented on vitrina/smart_contracts/views.py:
```diff
@@ -0,0 +1,124 @@
+            AgreementScope.objects.bulk_create(
+                [
+                    AgreementScope(
+                        agreement=agreement, resource=scope, action=scope.split("_")[-1]
```
`action=scope.split("_")[-1]` - why this part is needed?

martynasracct replied:
Example scope: `datasets_gov_vssa_rc_ar_person_getall` 

`AgreementScope.resource` saves full scope
`AgreementScope.action` only saves `getall` part

varzgalys commented on vitrina/smart_contracts/views.py:
```diff
@@ -0,0 +1,124 @@
+        return super().formset_valid(formset)
+
+    def formset_invalid(self, formset: BaseFormSet) -> HttpResponse:
+        messages.error(self.request, _("Sutartis generavime kilo klaidų"))
```
```suggestion
        messages.error(self.request, _("Sutarties generavime kilo klaidų"))
```

karina-klinkeviciute commented on tests/smart_contracts/test_views.py:
```diff
@@ -0,0 +1,165 @@
+        )
+        assert response.status_code == 404
+
+    def test_cannot_create_agreement_for_project_if_one_already_exists(
```
isn't it possible to add more agreements for the same project?

martynasracct replied:
The way it is implemented now - no. Maybe in future we should change it to always create new Agreements for newly added organizations

karina-klinkeviciute replied:
Dabar, manau, to užtenka. Ateityje turbūt reikėtų daryti, kad panaudojimo atvejis galėtų turėti daugiau nei vieną sutartį su tuo pačiu tiekėju. Nes visaip gali nutikti. Gal kas nors pradžioj nepagalvojo, kad jam reikės ir `search` funkcionalumo. Ar gal duomenų rinkinys pasipildė nauju modeliu, ar panašiai. Ir gali būti, kad to reikės tam pačiam panaudojimo atvejui.

karina-klinkeviciute commented on vitrina/smart_contracts/models.py:
```diff
@@ -19,3 +26,54 @@ class SmartContractTemplate(UUIDBaseModel):
+        on_delete=models.PROTECT,
+        verbose_name=_("Panaudojimo atvejis"),
+    )
+    organization = models.ForeignKey(
```
čia, kaip suprantu, yra duomenis teikianti organizacija? Ar gaunanti?

martynasracct replied:
I will make the name more specific

martynasracct replied:
organization is data provider organization. Changed the name to `assigner_organization`, so it is closer to ODRL

martynasracct replied:
Also, added better verbose name

karina-klinkeviciute commented on vitrina/smart_contracts/models.py:
```diff
@@ -19,3 +26,54 @@ class SmartContractTemplate(UUIDBaseModel):
+
+    class Meta:
+        managed = True
+        db_table = "usecase_agreement_scope"
```
čia, manau, irgi visai tinka numatytasis lentelės pavadinimas

martynasracct replied:
`Project` model is called `usecase` in database. By this logic, I called `Agreement` - `usecase_agreement` and `AgreementScope` - `usecase_agreement_scope`

martynasracct replied:
Removed custom DB table name

karina-klinkeviciute commented:
Kaip supratau, su šiuo PR ne visi punktai iš užduoties įgyvendinami?

martynasracct commented:
Taip. Šiuo PR įgyvendinami ne visi punktai

martynasracct committed:
1688 Agreement List and Details. Added AgreementFile model and admin

martynasracct committed:
1668 Fixes after QA testing

Changes "Generate agreement" texts to plural form.
Removes `Agreement.is_agent_sync_enabled` from UI.
Renames DB field `Agreement.assigner_organization` -> `Agreement.assigner`.
Removes field `Agreement.template` and adds `AgreementFile.is_template` instead. Assigned agreement template file will now be saved as any other Agreement file.
`AgreementScope.resource` saves only resource part of the scope (without action) and adds field `AgreementScope.scope` that saves full scope name (resource+action)
Fixes bug where datasets with empty name generated scopes without resource. Now scopes will not be generated in such cases

martynasracct committed:
Remove incorrectly added migration dependency
